### PR TITLE
[TRB-46436] Add USED value of the SEGMENTATION commodities sold by NODES

### PR DESCRIPTION
**Pull request by libai98**
# Intent

_Add USED value of the SEGMENTATION commodities sold by NODES_

# Background

   1. missing stitching metadata for SEGMENTATION commodities of PR https://github.com/turbonomic/kubeturbo/pull/919 not merged into this feature branch.
# Testing

_(any manual resting performed, along with log outputs and screenshots as appropriate)_

# Checklist

These are the items that must be done by the developer and by reviewers before the change is ready to merge. Please ~~strikeout~~ any items that are not applicable, but don't delete them

- [ ] Developer Checks
    - [X] Full build with unit tests and fmt and vet checks
    - [X] Unit tests added / updated
    - [X] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Integration tests added / updated
    - [ ] Manual testing done (and described)
    - [ ] Product sweep run and passed
    - [ ] Developer wiki updated (and linked to this description)
- [ ] Reviewer Checks
    - [ ] Merge request description clear and understandable
    - [ ] Developer checklist items complete
    - [ ] Functional code review (how is the code written)
    - [ ] Architectural review (does the code try to do the right thing, in the right way)
    - [ ] Defensive coding (incoming data checked / sanitized, exceptions logged, clear error messages)
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Security review checklist complete.

# Audience

_(@ mention any `review/...` groups or people that should be aware of this merge request)_

### CODE REVIEW

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 1]
```diff
@@ -444,10 +446,27 @@ func (builder *nodeEntityDTOBuilder) getAffinityCommoditiesSold(node *api.Node,
...
+		// Concatenate runningPods and pendingPods into a single slice and extract pod name
+		allPods := append(builder.clusterSummary.GetPendingPodsOnNode(node), builder.clusterSummary.GetRunningPodsOnNode(node)...)
+		for _, pod := range allPods {
+			podQualifiedNames.Insert(fmt.Sprintf("%s/%s", pod.Namespace, pod.Name))
+		}
+	}
+
+	var used float64
+	for workloadKey := range hostnameSpreadWorkloads {
+		workloadKeyPods := hostnameSpreadWorkloads[workloadKey]
```
Comment (kevinwangcn) [inline pkg/discovery/dtofactory/node_entity_dto_builder.go:460]: for workloadKey,workloadKeyPods:= range hostnameSpreadWorkloads {

Comment (libai98) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go]: Fixed. Thanks!

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 2]
```diff
@@ -422,7 +424,7 @@ func (builder *nodeEntityDTOBuilder) getAllocationCommoditiesSold(node *api.Node
 }
 
 func (builder *nodeEntityDTOBuilder) getAffinityCommoditiesSold(node *api.Node, nodesPods map[string][]string,
-	hostnameSpreadWorkloads sets.String, otherSpreadPods sets.String, podsToControllers map[string]string) []*proto.CommodityDTO {
+	hostnameSpreadWorkloads map[string]sets.String, otherSpreadPods sets.String, podsToControllers map[string]string) []*proto.CommodityDTO {
```
[the remark this answers is not in the document]
Comment (libai98) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go]: Yes, you are right.It is much simpler!

File: pkg/discovery/dtofactory/node_group_entity_dto_builder.go [hunk 3]
```diff
@@ -29,17 +30,29 @@ func NewNodeGroupEntityDTOBuilder(clusterSummary *repository.ClusterSummary, oth
...
-				nodeLst = sets.NewString(string(node.UID))
-			} else {
-				nodeLst.Insert(string(node.UID))
+
+			if _, exists := nodeGrp2nodes[fullLabel]; !exists {
+				nodeGrp2nodes[fullLabel] = sets.NewString()
+			}
+			nodeGrp2nodes[fullLabel].Insert(string(node.UID))
+
+			//add code for nodeGrp2workload
```
Comment (kevinwangcn) [inline pkg/discovery/dtofactory/node_group_entity_dto_builder.go:45]: All looks good to me except for this comment. Maybe simply remove it.

**Review by kevinwangcn — APPROVED** [review]

**Review by pankajbatra1 — APPROVED** [review]

File: pkg/discovery/dtofactory/node_group_entity_dto_builder.go [hunk 4]
```diff
@@ -29,17 +30,29 @@ func NewNodeGroupEntityDTOBuilder(clusterSummary *repository.ClusterSummary, oth
...
+			if _, exists := nodeGrp2workloads[fullLabel]; !exists {
+				nodeGrp2workloads[fullLabel] = sets.NewString()
+			}
+
+			for _, pod := range allPods {
+				podQualifiedName := fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+				if ctrlName, exists := builder.clusterSummary.PodToControllerMap[podQualifiedName]; exists {
+					nodeGrp2nodes[fullLabel].Insert(ctrlName)
```
[the remark this answers is not in the document]
Comment (libai98) [reply pkg/discovery/dtofactory/node_group_entity_dto_builder.go:53]: Yes, you are right! It should be nodeGrp2workloads. Thanks!

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 5]
```diff
@@ -444,10 +447,26 @@ func (builder *nodeEntityDTOBuilder) getAffinityCommoditiesSold(node *api.Node,
 		commoditiesSold = append(commoditiesSold, commodityDTO)
 	}
 
+	var allNodePodWorkloads sets.String
```
[the remark this answers is not in the document]
Comment (libai98) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go:450]: This is a name set of all the workload of running/pending pods on the node. It has nothing to do with spread.

File: pkg/discovery/dtofactory/node_group_entity_dto_builder.go (hunk 3 above)
Comment (libai98) [reply pkg/discovery/dtofactory/node_group_entity_dto_builder.go]: Removed. Thanks for the help for reviewing this.

File: pkg/discovery/dtofactory/node_entity_dto_builder.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (libai98) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go:450]: Maybe thisNodePodsWorkloads?

**Pull request merged**